src/runtime/virtcontainers/nydus.go

```diff
@@ -0,0 +1,349 @@
+	if err != nil {
+		return nil, err
+	}
+	return &NydusClient{
+		httpClient: &http.Client{
+			Timeout:   defaultHttpClientTimeout,
+			Transport: transport,
+		},
+	}, nil
+}
+
+func waitUntilSocketReady(sock string) error {
+	b := &backoff.ExponentialBackOff{
+		InitialInterval:     300 * time.Microsecond,
+		RandomizationFactor: 0.5,
+		Multiplier:          2,
+		MaxInterval:         1 * time.Second,
+		MaxElapsedTime:      5 * time.Second,
+		Clock:               backoff.SystemClock,
+	}
+	b.Reset()
+	err := backoff.Retry(func() error {
+		_, err := os.Stat(sock)
+		return err
+	}, b)
+	if err != nil {
+		return errors.Wrap(err, "nydusd daemon socket not ready")
+	}
+	return nil
+}
+
+func buildTransport(sock string) (http.RoundTripper, error) {
+	err := waitUntilSocketReady(sock)
+	if err != nil {
+		return nil, err
+	}
+	return &http.Transport{
+		MaxIdleConns:          10,
+		IdleConnTimeout:       10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
+			dialer := &net.Dialer{
+				Timeout:   5 * time.Second,
+				KeepAlive: 5 * time.Second,
+			}
+			return dialer.DialContext(ctx, "unix", sock)
+		},
+	}, nil
+}
+
+func (c *NydusClient) CheckStatus() (DaemonInfo, error) {
+	resp, err := c.httpClient.Get(fmt.Sprintf("http://unix%s", infoEndpoint))
+	if err != nil {
+		return DaemonInfo{}, err
+	}
+	defer resp.Body.Close()
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return DaemonInfo{}, err
+	}
+	var info DaemonInfo
+	if err = json.Unmarshal(b, &info); err != nil {
+		return DaemonInfo{}, err
+	}
+	return info, nil
+}
+
+func (c *NydusClient) Mount(mountPoint string, mr *MountRequest) error {
+	requestURL := fmt.Sprintf("http://unix%s?mountpoint=%s", mountEndpoint, mountPoint)
+	body, err := json.Marshal(mr)
+	if err != nil {
+		return errors.Wrap(err, "failed to create mount request")
+	}
+
+	resp, err := c.httpClient.Post(requestURL, contentType, bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNoContent {
+		return nil
+	}
+	return handleMountError(resp.Body)
+}
+
+func (c *NydusClient) Umount(sharedMountPoint string) error {
+	requestURL := fmt.Sprintf("http://unix%s?mountpoint=%s", mountEndpoint, sharedMountPoint)
+	req, err := http.NewRequest(http.MethodDelete, requestURL, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNoContent {
+		return nil
+	}
+	return handleMountError(resp.Body)
+}
+
+func handleMountError(r io.Reader) error {
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	var errMessage ErrorMessage
+	if err = json.Unmarshal(b, &errMessage); err != nil {
+		return err
+	}
+	return errors.New(errMessage.Message)
+}
+
+func parseConfigs(configs []string) (string, string, error) {
```

Comment:
please document the exact the config format, as this is an external ABI between snapshotter and kata runtime.